src/parser/daphnedsl/DaphneDSLVisitor.cpp

```diff
@@ -643,52 +644,137 @@ antlrcpp::Any DaphneDSLVisitor::visitParanthesesExpr(DaphneDSLGrammarParser::Par
-            auto specializedMatTy = argVal.getType().dyn_cast<mlir::daphne::MatrixType>();
-            bool isMatchingUnknownMatrix =
-                funcMatTy && specializedMatTy && funcMatTy.getElementType() == utils.unknownType;
-            if(funcInputType != argVal.getType() && !isMatchingUnknownMatrix && funcInputType != utils.unknownType) {
+            if (!argAndUDFParamCompatible(argVal.getType(), funcParamType)) {
                 compatible = false;
                 break;
             }
         }
         if (compatible) {
-            // TODO: variable results
-            return builder
-                .create<mlir::daphne::GenericCallOp>(loc,
-                    userDefinedFunc.sym_name(),
-                    args,
-                    userDefinedFunc.getType().getResults())
-                .getResult(0);
+            return userDefinedFunc;
+        }
+    }
+    // UDF with the provided name exists, but no version matches the argument types
+    if (range.second != range.first) {
+        // FIXME: disallow user-defined function with same name as builtins, otherwise this would be wrong behaviour
+        throw std::runtime_error("No function definition of `" + functionName + "` found with matching types");
+    }
+
+    // UDF with the provided name does not exist
+    return std::nullopt;
+}
+
+
+std::optional<mlir::FuncOp> DaphneDSLVisitor::findMatchingUnaryUDF(const std::string &functionName, mlir::Type argType) const {
+    // search user defined functions
+    auto range = functionsSymbolMap.equal_range(functionName);
+    
+    // TODO: find not only a matching version, but the `most` specialized    
+    for (auto it = range.first; it != range.second; ++it) {
+        auto userDefinedFunc = it->second;
+        auto funcTy = userDefinedFunc.getType();
+
+        if (funcTy.getNumInputs() != 1) {
+            continue;
+        }
+
+        if (argAndUDFParamCompatible(argType, funcTy.getInput(0))) {
+            return userDefinedFunc;
         }
     }
+    // UDF with the provided name exists, but no version matches the argument types
     if (range.second != range.first) {
         // FIXME: disallow user-defined function with same name as builtins, otherwise this would be wrong behaviour
-        throw std::runtime_error("No function definition of `" + func + "` found with matching types");
+        throw std::runtime_error("No function definition of `" + functionName + "` found with matching types");
+    }
+
+    // UDF with the provided name does not exist
+    return std::nullopt;
+}
+
+antlrcpp::Any DaphneDSLVisitor::handleMapOpCall(DaphneDSLGrammarParser::CallExprContext * ctx) {
+    std::string func = ctx->func->getText();
+    assert(func == "map" && "Called 'handleMapOpCall' for another function");
+    mlir::Location loc = utils.getLoc(ctx->start);
+    
+    if (ctx->expr().size() != 2) {
+      throw std::runtime_error(
```

Comment:
Note: The exception text is a copy of the exception thrown in `checkArgsExact`.

Reply:
Totally fine!